File: scripts/md_to_sql.py
#!/usr/bin/env python3
import os
import sqlite3
import json
import datetime
import re
import hashlib
from dotenv import load_dotenv

# --- CONFIGURATION ---
load_dotenv()

CONTENT_DIR = 'blogs'
OUTPUT_DIR = 'public/assets'
DB_NAME = 'db.sqlite'

# Auth Config
ADMIN_PASSWORD = os.getenv("ADMIN_PASSWORD", "123abc")
ADMIN_TOKEN = os.getenv("ADMIN_TOKEN", "ghp_mock_token")

# Deps
try:
    import frontmatter
    HAS_FRONTMATTER = True
except ImportError:
    HAS_FRONTMATTER = False

try:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
    from cryptography.hazmat.primitives import hashes
    HAS_CRYPTO = True
except ImportError:
    HAS_CRYPTO = False
    print("⚠️  'cryptography' module missing. Admin auth updates will be skipped.")

# --- UTILS ---

def ensure_dir(d):
    if not os.path.exists(d): os.makedirs(d)

def get_slug(path, root):
    rel = os.path.relpath(path, root)
    clean = os.path.splitext(rel)[0]
    slug = '/' + clean.replace(os.sep, '/')
    if slug.endswith('/index'): slug = slug[:-6]
    return slug

def parse_blocks(text):
    blocks = []
    lines = text.split('\n')
    lst = []
    
    def flush():
        if lst:
            blocks.append({ "type": "list", "data": { "style": "unordered", "items": lst[:] }})
            lst.clear()

    for line in lines:
        s = line.strip()
        if not s: flush(); continue
        if re.match(r'^(#{1,6})\s+(.*)', s):
            hm = re.match(r'^(#{1,6})\s+(.*)', s)
            flush()
            blocks.append({ "type": "header", "data": { "text": hm.group(2), "level": len(hm.group(1)) }})
            continue
        if re.match(r'^[\-\*]\s+(.*)', s):
            lm = re.match(r'^[\-\*]\s+(.*)', s)
            lst.append(lm.group(1).replace('**', '<b>').replace('**', '</b>'))
            continue
        flush()
        blocks.append({ "type": "paragraph", "data": { "text": s.replace('**', '<b>').replace('**', '</b>') }})
    flush()
    return json.dumps({ "time": int(datetime.datetime.now().timestamp()*1000), "blocks": blocks })

def fmt_date(d):
    if isinstance(d, (datetime.date, datetime.datetime)): return d.isoformat().split('T')[0]
    return str(d) if d else None

# --- AUTH ---

def update_admin_auth(conn):
    if not HAS_CRYPTO:
        return "Skipped (Missing 'cryptography' lib)"
    
    print("🔐 Syncing Admin Credentials from .env...")
    
    salt = os.urandom(16)
    kdf = PBKDF2HMAC(algorithm=hashes.SHA256(), length=32, salt=salt, iterations=100000)
    key = kdf.derive(ADMIN_PASSWORD.encode())
    key_hash = hashlib.sha256(key).hexdigest()
    
    iv = os.urandom(12)
    aesgcm = AESGCM(key)
    encrypted = aesgcm.encrypt(iv, ADMIN_TOKEN.encode(), None)
    
    c = conn.cursor()
    c.execute("DELETE FROM auth_store")
    c.execute("INSERT INTO auth_store VALUES (?, ?, ?, ?)", 
              (salt.hex(), key_hash, iv.hex(), encrypted.hex()))
    conn.commit()
    return "Synced with .env"

# --- DB INIT ---

def validate_schema(conn, table):
    try:
        c = conn.cursor()
        cols = [r[1] for r in c.execute(f"PRAGMA table_info({table})").fetchall()]
        if not cols: return
        
        # Check if we have 'date' column for posts, if not, drop to recreate
        required_col = 'date' if table == 'posts' else 'slug'
        if required_col not in cols:
            print(f"⚠️  Schema mismatch in '{table}' (Missing {required_col}). Recreating...")
            c.execute(f"DROP TABLE {table}")
    except: pass

def init_db(conn):
    conn.execute("CREATE TABLE IF NOT EXISTS auth_store (salt TEXT, key_hash TEXT, iv TEXT, encrypted_payload TEXT)")
    conn.execute("CREATE TABLE IF NOT EXISTS comments (id INTEGER PRIMARY KEY, post_id INTEGER, author TEXT, content TEXT, status TEXT)")
    conn.execute("CREATE TABLE IF NOT EXISTS v86_states (id INTEGER PRIMARY KEY, os_type TEXT, name TEXT, date TEXT, state_blob BLOB)")

    for t in ['posts', 'spec', 'resume']: validate_schema(conn, t)

    # UPDATED SCHEMA: 'published' -> 'date'
    conn.execute("""CREATE TABLE IF NOT EXISTS posts (
        id INTEGER PRIMARY KEY, slug TEXT UNIQUE, title TEXT, date TEXT, updated TEXT, 
        draft INTEGER DEFAULT 0, description TEXT DEFAULT "", image TEXT DEFAULT "", 
        tags TEXT DEFAULT "[]", category TEXT DEFAULT "", lang TEXT DEFAULT "", 
        media_type TEXT DEFAULT "text", media_url TEXT DEFAULT "",
        prevTitle TEXT DEFAULT "", prevSlug TEXT DEFAULT "", nextTitle TEXT DEFAULT "", nextSlug TEXT DEFAULT "", 
        content TEXT
    )""")
    conn.execute("CREATE TABLE IF NOT EXISTS spec (id INTEGER PRIMARY KEY, slug TEXT UNIQUE, title TEXT, content TEXT)")
    conn.execute("CREATE TABLE IF NOT EXISTS resume (id INTEGER PRIMARY KEY, slug TEXT UNIQUE, title TEXT, description TEXT DEFAULT '', published TEXT, draft INTEGER DEFAULT 0, content TEXT)")
    conn.commit()

def get_existing(conn, table):
    try: return set(r[0] for r in conn.execute(f"SELECT slug FROM {table}").fetchall())
    except: return set()

# --- MAIN ---

def main():
    ensure_dir(OUTPUT_DIR)
    db_path = os.path.join(OUTPUT_DIR, DB_NAME)
    print(f"🔌 Connecting to: {db_path}")
    conn = sqlite3.connect(db_path)
    init_db(conn)
    
    summary = []

    # 1. POSTS
    print("\n📂 Processing 'posts'...")
    existing = get_existing(conn, 'posts')
    added = 0
    skipped = 0
    new_items = []
    
    posts_dir = os.path.join(CONTENT_DIR, 'posts')
    if os.path.exists(posts_dir):
        for root, _, files in os.walk(posts_dir):
            for f in files:
                if not f.endswith('.md'): continue
                path = os.path.join(root, f)
                slug = get_slug(path, posts_dir)
                if slug in existing:
                    skipped += 1
                    continue
                
                post = frontmatter.load(path)
                meta = post.metadata
                new_items.append({
                    "slug": slug,
                    "title": meta.get('title', 'Untitled'),
                    "date": fmt_date(meta.get('date', meta.get('published'))), # Map published/date to date
                    "updated": fmt_date(meta.get('updated')),
                    "draft": 1 if meta.get('draft') else 0,
                    "description": meta.get('description', ''),
                    "image": meta.get('image', ''),
                    "tags": json.dumps(meta.get('tags', [])),
                    "category": meta.get('category', ''),
                    "lang": meta.get('lang', ''),
                    "media_type": meta.get('media_type', 'text'),
                    "media_url": meta.get('media_url', meta.get('image', '')),
                    "content": parse_blocks(post.content)
                })

        new_items.sort(key=lambda x: x['date'] or "", reverse=True)
        for i, item in enumerate(new_items):
            if i < len(new_items)-1: item['prevTitle'] = new_items[i+1]['title']; item['prevSlug'] = new_items[i+1]['slug']
            else: item['prevTitle'] = ""; item['prevSlug'] = ""
            if i > 0: item['nextTitle'] = new_items[i-1]['title']; item['nextSlug'] = new_items[i-1]['slug']
            else: item['nextTitle'] = ""; item['nextSlug'] = ""
            
            try:
                conn.execute("""INSERT INTO posts (slug, title, date, updated, draft, description, image, tags, category, lang, media_type, media_url, prevTitle, prevSlug, nextTitle, nextSlug, content) 
                             VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)""",
                             (item['slug'], item['title'], item['date'], item['updated'], item['draft'], 
                              item['description'], item['image'], item['tags'], item['category'], item['lang'], 
                              item['media_type'], item['media_url'],
                              item['prevTitle'], item['prevSlug'], item['nextTitle'], item['nextSlug'], item['content']))
                print(f"   [+] Added: {item['slug']}")
                added += 1
            except: pass
    summary.append(f"Posts:  {added} new, {skipped} skipped")

    # 2. OTHERS
    for name, fields in [('spec', ['title']), ('resume', ['title', 'description', 'published', 'draft'])]:
        print(f"\n📂 Processing '{name}'...")
        d = os.path.join(CONTENT_DIR, name)
        ex = get_existing(conn, name)
        add = 0; skp = 0
        if os.path.exists(d):
            for root, _, files in os.walk(d):
                for f in files:
                    if not f.endswith('.md'): continue
                    path = os.path.join(root, f)
                    slug = get_slug(path, d)
                    if slug in ex: skp+=1; continue
                    
                    post = frontmatter.load(path)
                    vals = [slug]
                    for field in fields:
                        val = post.metadata.get(field)
                        if field == 'tags': val = json.dumps(val or [])
                        elif field == 'draft': val = 1 if val else 0
                        elif 'date' in field or 'published' in field: val = fmt_date(val)
                        vals.append(val or "")
                    vals.append(parse_blocks(post.content))
                    
                    try:
                        cols = f"slug, {', '.join(fields)}, content"
                        ph = ",".join(["?"]*len(vals))
                        conn.execute(f"INSERT INTO {name} ({cols}) VALUES ({ph})", vals)
                        print(f"   [+] Added: {slug}")
                        add += 1
                    except: pass
        summary.append(f"{name.capitalize()}: {add} new, {skp} skipped")

    # 3. AUTH UPDATE
    print("\n🔑 Updating Admin Security...")
    auth_status = update_admin_auth(conn)
    summary.append(f"Auth:   {auth_status}")

    conn.commit()
    conn.close()

    print("\n" + "="*40)
    print("📊 DATABASE UPDATE SUMMARY")
    print("="*40)
    for line in summary: print(line)
    print("="*40 + "\n")

if __name__ == "__main__":
    main()

File: scripts/md_to_sql.js
require('dotenv').config();
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const matter = require('gray-matter');
const Database = require('better-sqlite3');

// --- CONFIGURATION ---
const CONTENT_DIR = 'blogs';
const OUTPUT_DIR = path.join('public', 'assets');
const DB_PATH = path.join(OUTPUT_DIR, 'db.sqlite');

// Auth
const ARGS = process.argv.slice(2);
const ADMIN_PASSWORD = process.env.ADMIN_PASSWORD || ARGS[0] || "123abc";
const ADMIN_TOKEN = process.env.ADMIN_TOKEN || "ghp_mock_token";

// --- UTILS ---

function ensureDir(dir) {
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
}

function getSlug(filePath, baseDir) {
    const rel = path.relative(baseDir, filePath);
    const parsed = path.parse(rel);
    let slug = '/' + path.join(parsed.dir, parsed.name).split(path.sep).join('/');
    if (slug.endsWith('/index')) slug = slug.substring(0, slug.length - 6);
    return slug;
}

function parseMdToBlocks(text) {
    const blocks = [];
    const lines = text.split('\n');
    let currentList = [];

    const flushList = () => {
        if (currentList.length > 0) {
            blocks.push({ type: "list", data: { style: "unordered", items: [...currentList] } });
            currentList = [];
        }
    };

    for (let line of lines) {
        const stripped = line.trim();
        if (!stripped) { flushList(); continue; }

        const headerMatch = stripped.match(/^(#{1,6})\s+(.*)/);
        if (headerMatch) {
            flushList();
            blocks.push({ type: "header", data: { text: headerMatch[2], level: headerMatch[1].length } });
            continue;
        }

        const listMatch = stripped.match(/^[\-\*]\s+(.*)/);
        if (listMatch) {
            currentList.push(listMatch[1].replace(/\*\*(.*?)\*\*/g, '<b>$1</b>'));
            continue;
        }

        flushList();
        blocks.push({ type: "paragraph", data: { text: stripped.replace(/\*\*(.*?)\*\*/g, '<b>$1</b>') } });
    }
    flushList();
    return JSON.stringify({ time: Date.now(), blocks: blocks });
}

function formatDate(dateObj) {
    if (!dateObj) return null;
    if (dateObj instanceof Date) return dateObj.toISOString().split('T')[0];
    return String(dateObj);
}

function* walkSync(dir) {
    if (!fs.existsSync(dir)) return;
    const files = fs.readdirSync(dir, { withFileTypes: true });
    for (const file of files) {
        if (file.isDirectory()) yield* walkSync(path.join(dir, file.name));
        else yield path.join(dir, file.name);
    }
}

// --- AUTH ---

function generateAuth(password, payload) {
    const salt = crypto.randomBytes(16);
    const key = crypto.pbkdf2Sync(password, salt, 100000, 32, 'sha256');
    const keyHash = crypto.createHash('sha256').update(key).digest('hex');
    const iv = crypto.randomBytes(12);
    const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
    let encrypted = cipher.update(payload, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    const authTag = cipher.getAuthTag().toString('hex');
    return { salt: salt.toString('hex'), keyHash, iv: iv.toString('hex'), encryptedPayload: encrypted + authTag };
}

// --- DB INIT ---

function validateAndMigrate(db, tableName) {
    try {
        const info = db.pragma(`table_info(${tableName})`);
        if (info.length === 0) return;
        const cols = info.map(c => c.name);

        // Ensure 'date' column exists for posts
        const required = tableName === 'posts' ? 'date' : 'slug';
        if (!cols.includes(required)) {
            console.log(`⚠️  Schema mismatch in '${tableName}' (Missing ${required}). Recreating...`);
            db.exec(`DROP TABLE ${tableName}`);
        }
    } catch (err) { console.error(err); }
}

function initDb(db) {
    db.exec("CREATE TABLE IF NOT EXISTS auth_store (salt TEXT, key_hash TEXT, iv TEXT, encrypted_payload TEXT)");
    db.exec("CREATE TABLE IF NOT EXISTS comments (id INTEGER PRIMARY KEY, post_id INTEGER, author TEXT, content TEXT, status TEXT)");
    db.exec("CREATE TABLE IF NOT EXISTS v86_states (id INTEGER PRIMARY KEY, os_type TEXT, name TEXT, date TEXT, state_blob BLOB)");

    ['posts', 'spec', 'resume'].forEach(t => validateAndMigrate(db, t));

    // UPDATED SCHEMA: 'published' -> 'date'
    db.exec(`
        CREATE TABLE IF NOT EXISTS posts (
            id INTEGER PRIMARY KEY, slug TEXT UNIQUE, title TEXT, date TEXT, updated TEXT,
            draft INTEGER DEFAULT 0, description TEXT DEFAULT "", image TEXT DEFAULT "",
            tags TEXT DEFAULT "[]", category TEXT DEFAULT "", lang TEXT DEFAULT "",
            media_type TEXT DEFAULT "text", media_url TEXT DEFAULT "",
            prevTitle TEXT DEFAULT "", prevSlug TEXT DEFAULT "", nextTitle TEXT DEFAULT "", nextSlug TEXT DEFAULT "",
            content TEXT
        )
    `);
    db.exec(`CREATE TABLE IF NOT EXISTS spec (id INTEGER PRIMARY KEY, slug TEXT UNIQUE, title TEXT, content TEXT)`);
    db.exec(`CREATE TABLE IF NOT EXISTS resume (id INTEGER PRIMARY KEY, slug TEXT UNIQUE, title TEXT, description TEXT DEFAULT "", published TEXT, draft INTEGER DEFAULT 0, content TEXT)`);
}

function updateAdminAuth(db) {
    console.log("🔐 Syncing Admin Credentials from .env...");
    const auth = generateAuth(ADMIN_PASSWORD, ADMIN_TOKEN);
    const txn = db.transaction(() => {
        db.exec("DELETE FROM auth_store");
        db.prepare("INSERT INTO auth_store VALUES (?, ?, ?, ?)").run(
            auth.salt, auth.keyHash, auth.iv, auth.encryptedPayload
        );
    });
    txn();
    return "Synced with .env";
}

function getExistingSlugs(db, tableName) {
    try {
        return new Set(db.prepare(`SELECT slug FROM ${tableName}`).all().map(r => r.slug));
    } catch { return new Set(); }
}

// --- PROCESSORS ---

function processPosts(db, rootPath) {
    const existingSlugs = getExistingSlugs(db, 'posts');
    const newItems = [];
    let added = 0, skipped = 0;

    console.log(`   Scanning ${rootPath}...`);
    for (const filePath of walkSync(rootPath)) {
        if (!filePath.endsWith('.md')) continue;
        const slug = getSlug(filePath, rootPath);
        if (existingSlugs.has(slug)) { skipped++; continue; }

        const { data, content } = matter(fs.readFileSync(filePath, 'utf8'));
        newItems.push({
            slug,
            title: data.title || 'Untitled',
            date: formatDate(data.date || data.published || new Date()),
            updated: formatDate(data.updated),
            draft: data.draft ? 1 : 0,
            description: data.description || '',
            image: data.image || '',
            tags: JSON.stringify(data.tags || []),
            category: data.category || '',
            lang: data.lang || '',
            media_type: data.media_type || 'text',
            media_url: data.media_url || data.image || '',
            contentRaw: content
        });
    }

    newItems.sort((a, b) => new Date(b.date) - new Date(a.date));

    for (let i = 0; i < newItems.length; i++) {
        if (i < newItems.length - 1) { newItems[i].prevTitle = newItems[i+1].title; newItems[i].prevSlug = newItems[i+1].slug; }
        if (i > 0) { newItems[i].nextTitle = newItems[i-1].title; newItems[i].nextSlug = newItems[i-1].slug; }
    }

    const insert = db.prepare(`INSERT INTO posts (slug, title, date, updated, draft, description, image, tags, category, lang, media_type, media_url, prevTitle, prevSlug, nextTitle, nextSlug, content) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`);

    const txn = db.transaction((items) => {
        for (const item of items) {
            try {
                insert.run(
                    item.slug, item.title, item.date, item.updated, item.draft,
                    item.description, item.image, item.tags, item.category, item.lang,
                    item.media_type, item.media_url,
                    item.prevTitle, item.prevSlug, item.nextTitle, item.nextSlug,
                    parseMdToBlocks(item.contentRaw)
                );
                console.log(`   [+] Added: ${item.slug}`);
                added++;
            } catch (e) { console.error(`   [!] Error ${item.slug}: ${e.message}`); }
        }
    });
    if (newItems.length) txn(newItems);
    return { added, skipped };
}

function processSimple(db, rootPath, tableName, fields) {
    const existing = getExistingSlugs(db, tableName);
    let added = 0, skipped = 0;

    console.log(`   Scanning ${rootPath}...`);
    const stmt = db.prepare(`INSERT INTO ${tableName} (slug, ${fields.join(', ')}, content) VALUES (?, ${fields.map(()=>'?').join(', ')}, ?)`);

    const txn = db.transaction(() => {
        for (const filePath of walkSync(rootPath)) {
            if (!filePath.endsWith('.md')) continue;
            const slug = getSlug(filePath, rootPath);
            if (existing.has(slug)) { skipped++; continue; }

            const { data, content } = matter(fs.readFileSync(filePath, 'utf8'));
            const values = [slug];
            for (const f of fields) {
                let v = data[f];
                if (f==='tags') v = JSON.stringify(v||[]);
                else if (f==='draft') v = v?1:0;
                else if (f==='published'||f.includes('date')) v = formatDate(v);
                values.push(v || "");
            }
            values.push(parseMdToBlocks(content));

            try { stmt.run(...values); added++; console.log(`   [+] Added: ${slug}`); }
            catch (e) { console.error(e.message); }
        }
    });
    txn();
    return { added, skipped };
}

function main() {
    ensureDir(OUTPUT_DIR);
    console.log(`🔌 Connecting to database: ${DB_PATH}`);
    const db = new Database(DB_PATH);
    initDb(db);

    const summary = [];

    console.log("\n📂 Processing 'posts'...");
    const p = processPosts(db, path.join(CONTENT_DIR, 'posts'));
    summary.push(`Posts:  ${p.added} new, ${p.skipped} skipped`);

    console.log("\n📂 Processing 'spec'...");
    const s = processSimple(db, path.join(CONTENT_DIR, 'spec'), 'spec', ['title']);
    summary.push(`Spec:   ${s.added} new, ${s.skipped} skipped`);

    console.log("\n📂 Processing 'resume'...");
    const r = processSimple(db, path.join(CONTENT_DIR, 'resume'), 'resume', ['title', 'description', 'published', 'draft']);
    summary.push(`Resume: ${r.added} new, ${r.skipped} skipped`);

    console.log("\n🔑 Updating Admin Security...");
    const authStatus = updateAdminAuth(db);
    summary.push(`Auth:   ${authStatus}`);

    db.close();

    console.log("\n" + "=".repeat(40));
    console.log("📊 DATABASE UPDATE SUMMARY");
    console.log("=".repeat(40));
    summary.forEach(line => console.log(line));
    console.log("=".repeat(40) + "\n");
}

main();
